File: testEndlessScroll/testEndlessScroll/CollectionSectionDatasource.m
//
// Created by Victor Zinets on 2/6/18.
// Copyright (c) 2018 Together Networks. All rights reserved.
//

#import "CollectionSectionDatasource.h"

@implementation CollectionSectionDatasource {

}

- (void)setCollectionView:(UICollectionView *)collectionView {
    _collectionView = collectionView;
    _collectionView.dataSource = self;
}

- (void)setItems:(NSArray *)items {
    [super setItems:items];

    [self.collectionView performBatchUpdates:^{
        if (toRemove.count) {
            NSMutableArray *array = [NSMutableArray arrayWithCapacity:toRemove.count];
            [toRemove enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
                // todo секции! хз; может просто как свойство класса?
                [array addObject:[NSIndexPath indexPathForItem:idx inSection:0]];
            }];
            [self.collectionView deleteItemsAtIndexPaths:array];
        }
        if (toInsert.count) {
            NSMutableArray *array = [NSMutableArray arrayWithCapacity:toInsert.count];
            [toInsert enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
                [array addObject:[NSIndexPath indexPathForItem:idx inSection:0]];
            }];
            [self.collectionView insertItemsAtIndexPaths:array];
        }
        if (toUpdate.count) {
            NSMutableArray *array = [NSMutableArray arrayWithCapacity:toUpdate.count];
            [toUpdate enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
                [array addObject:[NSIndexPath indexPathForItem:idx inSection:0]];
            }];
            [self.collectionView reloadItemsAtIndexPaths:array];
        }
    } completion:^(BOOL finished) {

    }];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 0;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    return nil;
}

@end

File: yaScroller2/yaScroller2/classes/base datasource/CollectionSectionDatasource.swift
//
//  CollectionSectionDatasource.swift
//  yaScroller2
//
//  Created by Victor Zinets on 8/28/18.
//  Copyright © 2018 Victor Zinets. All rights reserved.
//

import UIKit

class CollectionSectionDatasource : SectionDatasource, UICollectionViewDataSource {
    var section: Int = 0
    var collectionView: UICollectionView? {
        didSet {
            if let collection = collectionView {
                collection.dataSource = self
            }
        }
    }
    
    // MARK: overrides -
    
    override var items:[AnyHashable] {
        get {
            return super.items
        }
        set (newItems) {
            super.items = newItems
            
            if let collection = collectionView {
                collection.performBatchUpdates({
                    if (toRemove.count > 0) {
                        var array = [IndexPath]()
                        for index in toRemove {
                            array.append(IndexPath(item: index, section: section))
                        }
                        collection.deleteItems(at: array)
                    }
                    if (!toInsert.isEmpty) {
                        var array = [IndexPath]()
                        for index in toInsert {
                            array.append(IndexPath(item: index, section: section))
                        }
                        collection.insertItems(at: array)
                    }
                    if (!toUpdate.isEmpty) {
                        var array = [IndexPath]()
                        for item in toUpdate {
                            array.append(IndexPath(item: item, section: section))
                        }
                        collection.reloadItems(at: array)
                    }
                }) { (finished) in
                    
                }
            }
        }
    }
    
    // MARK: collection support -
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 0
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }
    
}
